File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:

        def dfs(root, subRoot):
            if not root:
                return False
            if root.val == subRoot.val and check(root, subRoot):
                return True
            
            return dfs(root.left, subRoot) or dfs(root.right, subRoot)
        
        def check(root, subRoot):
            if not root and not subRoot:
                return True
            if not root or not subRoot:
                return False
            return check(root.left, subRoot.left) and check(root.right, subRoot.right) and root.val == subRoot.val
        
        return dfs(root, subRoot)
            

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {number[]} preorder
     * @param {number[]} inorder
     * @return {TreeNode}
     */
    buildTree(preorder, inorder) {
        if (!preorder.length || !inorder.length) {
            return null
        }
        let root = new TreeNode(preorder[0]);
        let mid = inorder.indexOf(root.val); 
        // Split the inorder and preorder arrays for left and right subtrees.
        let inorderLeft = inorder.slice(0, mid);
        let inorderRight = inorder.slice(mid + 1);

        let preorderLeft = preorder.slice(1, mid + 1);
        let preorderRight = preorder.slice(mid + 1);

        
        root.left = this.buildTree(preorderLeft, inorderLeft);
        root.right = this.buildTree(preorderRight, inorderRight);
        return root;
    }
}
